src/tools/linkchecker/main.rs

```diff
@@ -45,74 +47,262 @@ fn main() {
     let mut errors = false;
-    walk(&docs, &docs, &mut url, &mut errors);
+    walk(&mut HashMap::new(), &docs, &docs, &mut url, &mut errors);
     if errors {
         panic!("found some broken links");
     }
 }
 
-fn walk(root: &Path, dir: &Path, url: &mut Url, errors: &mut bool) {
+#[derive(Debug)]
+pub enum LoadError {
+    IOError(std::io::Error),
+    BrokenRedirect(PathBuf, std::io::Error),
+}
+
+struct FileEntry {
+    source: String,
+    ids: HashSet<String>,
+}
+
+type Cache = HashMap<PathBuf, FileEntry>;
+
+impl FileEntry {
+    fn parse_ids(&mut self,
+                file: &Path,
+                contents: &str,
+                errors: &mut bool)
+{
+        if self.ids.is_empty() {
+            with_attrs_in_source(contents, " id", |fragment, i| {
+                let frag = fragment.trim_left_matches("#").to_owned();
+                if !self.ids.insert(frag) {
+                    *errors = true;
+                    println!("{}:{}: id is not unique: `{}`",
+                             file.display(), i, fragment);
+                }
+            });
+        }
+    }
+}
+
+fn walk(cache: &mut Cache,
+        root: &Path,
+        dir: &Path,
+        url: &mut Url,
+        errors: &mut bool)
+{
     for entry in t!(dir.read_dir()).map(|e| t!(e)) {
         let path = entry.path();
         let kind = t!(entry.file_type());
         url.path_mut().unwrap().push(entry.file_name().into_string().unwrap());
         if kind.is_dir() {
-            walk(root, &path, url, errors);
+            walk(cache, root, &path, url, errors);
         } else {
-            check(root, &path, url, errors);
+            let pretty_path = check(cache, root, &path, url, errors);
+            if let Some(pretty_path) = pretty_path {
+                let entry = cache.get_mut(&pretty_path).unwrap();
+                // we don't need the source anymore,
+                // so drop to to reduce memory-usage
+                entry.source = String::new();
+            }
         }
         url.path_mut().unwrap().pop();
     }
 }
 
-fn check(root: &Path, file: &Path, base: &Url, errors: &mut bool) {
+fn check(cache: &mut Cache,
+         root: &Path,
+         file: &Path,
+         base: &Url,
+         errors: &mut bool) -> Option<PathBuf>
+{
     // ignore js files as they are not prone to errors as the rest of the
     // documentation is and they otherwise bring up false positives.
     if file.extension().and_then(|s| s.to_str()) == Some("js") {
-        return
+        return None;
     }
 
-    let pretty_file = file.strip_prefix(root).unwrap_or(file);
-
     // Unfortunately we're not 100% full of valid links today to we need a few
     // whitelists to get this past `make check` today.
     // FIXME(#32129)
-    if file.ends_with("std/string/struct.String.html") {
-        return
+    if file.ends_with("std/string/struct.String.html") ||
+       file.ends_with("collections/string/struct.String.html") {
```

Comment:
For the addition of these new ignores, can you be sure that issues are filed and these are tagged with a FIXME?